add support for A128CBC+HS256

msft-mw opened the pull request:
rolling our own crypto, what could go wrong

Commit:
hacks upon hacks

Commit:
clean up support for A128CBC+HS256

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -23,15 +23,95 @@ import (
+	return hashed[:len(key)]
+}
+
+// ComputeEncryptionKey derives an encryption key based on a master key, using the A128CBC+HS256 draft specification
```
These could actually be made a little more generic to support A128CBC+HS384/512 too. You can take a hash.Hash as an argument and use that instead of calling `sha256.Sum256`. Everything else is portable.

msft-mw replied:
For future, yes let's do this

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -23,15 +23,95 @@ import (
+// ComputeEncryptionKey derives an encryption key based on a master key, using the A128CBC+HS256 draft specification
+func ComputeEncryptionKey(key []byte, algorithm string) []byte {
+	keySize := len(key) / 2
+	label := "Encryption"
```
Encryption/Integrity could probably be made consts

msft-mw replied:
Done

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -23,15 +23,95 @@ import (
+// NewCBCHMACEx instantiates a new AEAD based on CBC+HMAC from the draft 7 spec
+func NewCBCHMACEx(key []byte, newBlockCipher func([]byte) (cipher.Block, error)) (cipher.AEAD, error) {
+	keySize := len(key) / 2
+	enc := "A128CBC+HS256"
```
Consider taking this as an argument?

msft-mw replied:
Sure

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -98,12 +178,67 @@ func (ctx *cbcAEAD) Seal(dst, nonce, plaintext, data []byte) []byte {
+// nonce = iv
+// ciphertext = cdata
+// adata = data
+// kdata = doesn't exist yet
```
These debugs can be 🔥 'd now

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -98,12 +178,67 @@ func (ctx *cbcAEAD) Seal(dst, nonce, plaintext, data []byte) []byte {
+		return nil, errors.New("square/go-jose: invalid ciphertext (too short)")
+	}
+
+	offset := len(ciphertext) - ctx.authtagBytes - 256 // sha256 digest size
```
The 256 in sha256 is 256 bits, not bytes. How does this work?

msft-mw replied:
Thanks, brain wasn't working. You got me. The bytes do seem to be fixed at 256, but it's not a SHA digest.

msft-mw replied:
This is actually the keysize in bytes; XSTS uses RSA-2048. Still a magic number, but now we know why.

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -98,12 +178,67 @@ func (ctx *cbcAEAD) Seal(dst, nonce, plaintext, data []byte) []byte {
+	dotPos := len(ciphertext) - 256 // length of a sha256 digest
+	// the inclusion of a dot after the ciphertext and auth tag indicates that there was additional data included
+	// in the buffer that needs to be handled and teased out
+	if subtle.ConstantTimeCompare(dot, ciphertext[dotPos-1:dotPos]) == 1 {
```
This can panic if the ciphertext is too short, don't think it's been verified at this point yet...

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -135,6 +270,23 @@ func (ctx *cbcAEAD) Open(dst, nonce, ciphertext, data []byte) ([]byte, error) {
+func (ctx *cbcAEAD) computeAuthTagEx(aad, kad, iv, ciphertext []byte) []byte {
+	buffer := append([]byte{}, aad...)
+	buffer = append(buffer, dot...)
+	blargh := []byte(base64URLEncode(kad))
```
Debug var name 😛

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -194,3 +346,22 @@ func unpadBuffer(buffer []byte, blockSize int) ([]byte, error) {
+const base64Character63 = "/"
+
+const base64UrlCharacter62 = "-"
+const base64UrlCharacter63 = "_"
```
Prefer the const block

```
const (
  base64...
  base64...
)
```

msft-mw replied:
Resolved

connor4312 commented on cipher/cbc_hmac.go:
```diff
@@ -194,3 +346,22 @@ func unpadBuffer(buffer []byte, blockSize int) ([]byte, error) {
+	// encode as base64
+	encoded := base64.StdEncoding.EncodeToString(input)
+
+	encoded = strings.Split(encoded, base64PadCharacter)[0]
```
You should be able to use RawStdEncoding instead, which won't add padding

msft-mw replied:
Resolved

connor4312 commented:
Since we ideally want to contribute this back upstream, I would look at trying to extract some logic so that we can avoid special casing and build more generic things. Also, you should be able to support A192CBC+HS384 and A256CBC+HS512 with a few parameter tweaks. Could be a nice addition.

Commit:
add unit test for testing xtoken cracking

Commit:
cleaning up based on PR feedback
- this is highly dependent on the fact that we use RSA-2048
- if it ever changes, or we support arbitrary key strengths, we'll be in hot water

connor4312 committed:
Allow cek/cik generation to support multiple hash sizes

connor4312 committed:
Support other kdfcbc algorithms, dynamic key size

connor4312 committed:
Remove unusde consts, simplify url encoding

connor4312 committed:
Address PR comments

connor4312 committed:
Merge pull request #2 from WatchBeam/generic

Generic

connor4312 commented:
Going to go ahead and get this in.
